### ISSUE TITLE
Trying to close a never connected websocket returns an endless future

### ISSUE THREAD

**Issue by Bungeefan** [dart-lang/web_socket_channel#231]
### Use case:
We tried to list to a websocket stream after we created it.
The websocket can't reach the server and fails with and error in the `onError` handler.

Finally when trying to close such a websocket with `await webSocket?.sink.close();`, it hangs in an endless future.

### Research:
As far my research went, I found the `done` getter in the `_CompleterSink` to be the underlying problem.

The following code returns our endless future:
https://github.com/dart-lang/web_socket_channel/blob/master/lib/src/sink_completer.dart#L76
```dart
@override
Future get done {
  if (_doneCompleter != null) return _doneCompleter!.future;
  if (_destinationSink == null) {
    _doneCompleter = Completer.sync();
    return _doneCompleter!.future;
  }
  return _destinationSink!.done;
}
```

It returns a future which waits for the destination sink, which only gets complete when `_setDestinationSink` is called.
Which in our case never gets calles, because the destination sink was never successfully created, therefore the future never completes.
Call that would complete the future, but only after successful creation: https://github.com/dart-lang/web_socket_channel/blob/master/lib/io.dart#L73

#### (First idea) Workaround:
I successfully tested the following code example which seems to fix it, however I am not sure if there are any possible sideeffects with this "solution".

This checks in the close method if there is any current destination sink, if there isn't the caller probably isn't interested in waiting for one to complete. Otherwise behave as before.

```dart
@override
Future close([int? closeCode, String? closeReason]) {
  if (_canSendDirectly) {
    _destinationSink!.close(closeCode, closeReason);
  } else {
    _closeCode = closeCode;
    _closeReason = closeReason;
    _ensureController().close();
  }
  if (_destinationSink == null) {
    return Future.value();
  }
  return done;
}
```

I also found another issue (https://github.com/dart-lang/web_socket_channel/issues/185) which seems to run in the same or a similar issue and the only comment was to not await the future, which isn't really a solution.

**Comment by wangdenkun** [comment]
this issues waste my hours today again. 😒

**Comment by anggaaryas** [comment]
no Update?

**Comment by JemmyWang94** [comment]
I also encountered this problem, is there any update?

**Comment by AD1N1993** [comment]
Is there any update? 2024!!!!!!!

**Comment by AD1N1993** [comment]
https://github.com/dart-lang/web_socket_channel/issues/231

**Comment by leeyisoft** [comment]
I have a feeling that my code will solve your problem, why not give it a try?

我感觉我这里代码能够解决你的问题，何不尝试一下呢？ 

@AD1N1993  @Bungeefan 

https://github.com/imboy-pub/imboy-flutter/issues/6
